util/unifyfs-shell/src/unifyfs-shell.c

```diff
@@ -0,0 +1,295 @@
+#include <limits.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <mpi.h>
+#include <unifyfs.h>
+
+#include "unifyfs-shell.h"
+
+static unifyfs_shell_cmd_t* commands[] = {
+    &unifyfs_shell_cmd_status,
+    &unifyfs_shell_cmd_ls,
+    &unifyfs_shell_cmd_mkdir,
+    &unifyfs_shell_cmd_cd,
+    &unifyfs_shell_cmd_pwd,
+    &unifyfs_shell_cmd_stat,
+};
+
+static unifyfs_shell_env_t* env;
+
+static int target;
+static char cmdbuf[LINE_MAX];
+
+#define UNIFYFS_SHELL_MAX_ARGC  16
+
+static int shell_argc;
+static char* shell_argv[UNIFYFS_SHELL_MAX_ARGC];
+
+static int parse_input(char* input)
+{
+    int ret = 0;
+    char* pos = input;
+
+    shell_argc = 0;
+
+     while (1) {
+        if (shell_argc >= UNIFYFS_SHELL_MAX_ARGC) {
+            ret = -E2BIG;
+            goto out;
+        }
+
+        while (isspace(pos[0])) {
+            pos++;
+        }
+
+        shell_argv[shell_argc++] = pos;
+
+        while (pos[0] != '\0' && !isspace(pos[0])) {
+            pos++;
+        }
+
+        if (pos[0] == '\0') {
+            break;
+        }
+
+        pos[0] = '\0';
+        pos++;
+    }
+
+out:
+     return ret;
+}
+
+static inline unifyfs_shell_cmd_t* find_command(const char* str)
+{
+    int i = 0;
+    int n_commands = sizeof(commands)/sizeof(unifyfs_shell_cmd_t*);
+    unifyfs_shell_cmd_t* cmd = NULL;
+
+    for (i = 0; i < n_commands; i++) {
+        if (0 == strcmp(commands[i]->name, str)) {
+            cmd = commands[i];
+            break;
+        }
+    }
+
+    return cmd;
+}
+
+static int execute_command(char* input, FILE* out)
+{
+    int ret = 0;
+    unifyfs_shell_cmd_t* cmd = NULL;
+
+    ret = parse_input(input);
+    if (ret < 0) {
+        fprintf(out, "too many arguments\n");
+        goto out;
+    }
+
+    env->output = out;
+    optind = 1;  /* for possible getopt */
+
+    cmd = find_command(shell_argv[0]);
+    if (cmd) {
+        ret = cmd->func(shell_argc, shell_argv, env);
+    } else {
+        fprintf(out, "Invalid command: %s\n", input);
+    }
+
+out:
+    fputc('\0', out);
+    fflush(out);
+
+    return ret;
+}
+
+static inline char* strtrim(char* str)
+{
+    char* pos = str;
+    char* rpos = &str[strlen(str) - 1];
+
+    if (pos[0] == '\0') {
+        return NULL;
+    }
+
+    while (isspace(pos[0])) {
+        pos++;
+
+        if (pos[0] == '\0') {
+            return NULL;
+        }
+    }
+
+    while (isspace(rpos[0]) && pos < rpos) {
+        rpos[0] = '\0';
+        rpos--;
+    }
+
+    return pos;
+}
+
+static char* shell_prompt(void)
+{
+    char* input = NULL;
+
+    do {
+        printf("[%d@unifyfs-shell]: ", target);
+        fflush(stdout);
+
+        input = strtrim(fgets(cmdbuf, LINE_MAX-1, stdin));
+    } while (!input);
+
+    return input;
+}
+
+static inline void unifyfs_shell_help(void)
+{
+    int i = 0;
+    int n_commands = sizeof(commands)/sizeof(unifyfs_shell_cmd_t*);
+
+    printf(
+      "Builtin commands:\n"
+      "  .help           this help message\n"
```

Comment:
any particular reason for prefixing the builtin commands with '.'?

Reply:
@MichaelBrim This was just to differentiate these meta commands from main commands that resemble binutil commands. I will remove the dot prefix if this is not agreeable.

Reply:
It seems weird to me - I think removing the dot prefix would be preferable. Another suggestion would be to create an 'exit' alias for the 'quit' builtin.

Reply:
@MichaelBrim Okay, I will remove the dot prefix.